server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -695,6 +708,38 @@ private int getCount(MNode node, String[] nodes, int idx) throws MetadataExcepti
     }
   }
 
+  /**
+   * Traverse the MTree to get the count of devices.
+   */
+  private int getDevicesCount(MNode node, String[] nodes, int idx) throws MetadataException {
+    String nodeReg = MetaUtils.getNodeRegByIdx(idx, nodes);
+    int cnt = 0;
+    if (!(PATH_WILDCARD).equals(nodeReg)) {
+      if (node.hasChild(nodeReg)) {
+        if (node.getChild(nodeReg) instanceof MeasurementMNode && idx >= nodes.length) {
+          cnt++;
+        } else {
+          cnt += getDevicesCount(node.getChild(nodeReg), nodes, idx + 1);
+        }
+      }
+    } else {
+      boolean deviceAdded = false;
+      for (MNode child : node.getChildren().values()) {
+        if (child instanceof MeasurementMNode) {
+          if (!deviceAdded) {
+            cnt++;
+            deviceAdded = true;
+          } else {
+            break;
+          }
+        } else {
+          cnt += getDevicesCount(child, nodes, idx + 1);
+        }
+      }
```

Comment:
Hi, we support subdevice like "root.sg.device1.subdevice1.sensor1" while subdevice1 is a measurementNode and a device. In your case, the subdevice1 will be ignored.

You can refer to the Mtree.findDevice() method.

Reply:
Ah.. good point. Maybe @yhwang-hbl could add a specific test case for this example ; )

Reply:
> Hi, we support subdevice like "root.sg.device1.subdevice1.sensor1" while subdevice1 is a measurementNode and a device. In your case, the subdevice1 will be ignored.
> 
> You can refer to the Mtree.findDevice() method.
hi, I didn't know the concept of sub devices before. I did a self-test on my code, which will filter out the situation you said. I will improve it,thanks for your guidance

Reply:
> Ah.. good point. Maybe @yhwang-hbl could add a specific test case for this example ; )

OK，i will improve it and add a test for the corresponding situation